// globi-hairball/lib/transform.js
module.exports = {
    parseInteractions: parseInteractions,
    toTaxon: toTaxon
};

function parseInteractions(data) {
    var nodes = [];
    var links = [];
    var nodeCache = {}, index = 0, interaction, source, target, sourceId, targetId, linkId, linkCache = {}, i;
    for (i = 0; interaction = data[i]; i++) {
        source = interaction['source'];
        source['id'] = source['id'].replace(':', '_');
        sourceId = source['id'];
        target = interaction['target'];
        target['id'] = target['id'].replace(':', '_');
        targetId = target['id'];

        if ((sourceId !== 'no_match') && !nodeCache[sourceId]) {
            nodeCache[sourceId] = source;
            nodeCache[sourceId]['index'] = index++;
        }
        if ((targetId !== 'no_match') && !nodeCache[targetId]) {
            nodeCache[targetId] = target;
            nodeCache[targetId]['index'] = index++;
        }

        if (
            (sourceId !== 'no_match') &&
                (targetId !== 'no_match') &&
                (sourceId !== targetId)
            ) {
            linkId = sourceId + '---' + targetId;
            if (!linkCache[linkId]) {
                linkCache[linkId] = {
                    source: nodeCache[sourceId]['index'],
                    target: nodeCache[targetId]['index'],
                    link_id: linkId
                };
            }
        }
    }

    for (var link in linkCache) {
        if (linkCache.hasOwnProperty(link)) {
            links.push(linkCache[link]);
        }
    }

    for (var node in nodeCache) {
        if (nodeCache.hasOwnProperty(node)) {
            nodes.push(nodeCache[node]);
        }
    }
    return { nodes: nodes, links: links};
}

function toTaxon(node) {
    return {id: node.id, name: node.name, path: node.path};
}
