// globi-spatial-selector/index.js
var fs = require('fs');
var insertCss = require('insert-css');
var inherits = require('inherits');
var EventEmitter = require('events').EventEmitter;
var AreaPicker = require('./lib/areapicker.js');
var infobox = require('./lib/infobox.js');
var boundsToBBox = require('./lib/boundsToBBox.js');
var d3 = require('globi').d3;

inherits(SpatialSelector, EventEmitter);
module.exports = SpatialSelector;

var currentInfoWindow;

function SpatialSelector(opts) {
    if (!(this instanceof SpatialSelector)) return new SpatialSelector(opts);
    this.opts = opts;
};


SpatialSelector.prototype.appendTo = function (target) {
    if (window.google) {
        if (typeof target === 'string') target = document.querySelector(target);
        var css = fs.readFileSync(__dirname + '/style.css', 'utf8');
        insertCss(css);
        this.appendMap(target, this.opts);
        this.emit('append', target);
    } else {
        this.emit('error', target);
    }
};

SpatialSelector.prototype.appendMarkersTo = function (json, map) {
    var markers = [];
    var locations = json.data;
    for (var index in locations) {
        if (locations.hasOwnProperty(index)) {
            var location = { latitude: locations[ index ][ 0 ], longitude: locations[ index ][ 1 ] },
                content = createLocationContent(location);
            markers.push(placeMarker(content, location, map));
        }
    }
    initializeMarkerClusterer(map, markers);
};

SpatialSelector.prototype.appendAreaPicker = function (map, params, google) {
    var picker = new AreaPicker(map, google);
    if (params.bbox) {
        picker.setBounds(toLatLngBounds(params.bbox, google)).show();
        picker.control_.setActive();
        map.panToBounds(picker.bounds_);
        this.emit('change', boundsToBBox(picker.bounds_));
    }
    return picker;
};

function toLatLngBounds(bbox, google) {
    return new google.maps.LatLngBounds(
        new google.maps.LatLng(bbox[ 1 ], bbox[ 0 ]),
        new google.maps.LatLng(bbox[ 3 ], bbox[ 2 ])
    );
}

SpatialSelector.prototype.appendMap = function (target, params) {
    var me = this;
    var google = window.google;
    var location = location || { latitude: 0, longitude: 0 };
    var zoom = zoom || 1;
    var options = {
        zoom: zoom,
        center: new google.maps.LatLng(location.latitude, location.longitude),
        mapTypeId: google.maps.MapTypeId.ROADMAP};
    var map = new google.maps.Map(target, options);
    me.appendAreaPicker(map, params, google);
    d3.json('http://api.globalbioticinteractions.org/locations', function (json) {
        if (json) {
            me.appendMarkersTo(json, map);
        }
    });
};

function initializeMarkerClusterer(map, markers) {
    return new MarkerClusterer(map, markers, { gridSize: 40, maxZoom: 0 });
}

function createLocationContent(location) {
    return infobox.locationInfoBox({location: {lng: location.longitude, lat: location.latitude }});
}

function placeMarker(content, location, map) {
    var latLng = new google.maps.LatLng(location.latitude, location.longitude),
        marker = new google.maps.Marker({ position: latLng, map: map }),
        infoWindow = new google.maps.InfoWindow({ content: content, maxWidth: 200 });

    google.maps.event.addListener(marker, 'click', function () {
        currentInfoWindow && currentInfoWindow.close();
        currentInfoWindow = infoWindow;
        currentInfoWindow.open(map, marker);
    });

    return marker;
}
